src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpOnTypeFormattingPass.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class CSharpOnTypeFormattingPass : FormattingPassBase
+    {
+        private readonly ILogger _logger;
+
+        public CSharpOnTypeFormattingPass(
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ILanguageServer server,
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<CSharpOnTypeFormattingPass>();
+        }
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result)
+        {
+            if (!context.IsFormatOnType || result.Kind != RazorLanguageKind.CSharp)
+            {
+                // We don't want to handle regular formatting or non-C# on type formatting here.
+                return result;
+            }
+
+            // Normalize and re-map the C# edits.
+            var codeDocument = context.CodeDocument;
+            var csharpText = SourceText.From(codeDocument.GetCSharpDocument().GeneratedCode);
+            var normalizedEdits = NormalizeTextEdits(csharpText, result.Edits);
+            var tempText = csharpText.WithChanges(normalizedEdits.Select(c => c.AsTextChange(csharpText)));
+            var mappedEdits = RemapTextEdits(codeDocument, normalizedEdits, result.Kind);
+            if (mappedEdits.Length == 0)
+            {
+                // There are no CSharp edits for us to apply. No op.
+                return new FormattingResult(mappedEdits);
+            }
+
+            // Find the lines that were affected by these edits.
+            var originalText = codeDocument.GetSourceText();
+            var changes = mappedEdits.Select(e => e.AsTextChange(originalText));
```

Comment:
Should we  be representing our edits as TextChanges  from  the  get-go? Wondering because it  seems like  we  do AsTextChange  everywhere

Reply:
Nope. There are a bunch of cases where we need the line/character information. We also use `AsRange` a lot.